README.md
# CoreMedia Monitoring Toolbox

#### Table of Contents

1. [About][About]
2. [Installation][Installation]
3. [Publications][publications]
4. [License][License]
5. [Contributing][Contributing]


## About

The Monitoring Toolbox is a set of monitoring tools that monitors the availability of CoreMedia resources and can generate performance data for reporting.
This data is made available as graphs and congestion changes.

It provides a simple web frontend to manage hosts.

<img alt="landing page" src="doc/assets/monitoring-toolbox_1.png" title="Landing Page" style="display:block;margin: 0 auto;width: 1024px">


## Installation

For installing please check the [Quick start](doc/de/02-getting-started.md) in the documentation.
Sorry, currently only available in german.


## Publications

Officially the stack was presented at the [OSMC](https://osmc.de/archive-2017/) (the worldwide leading conference focused on open source monitoring solutions).

The talk was published on [YouTube](https://www.youtube.com/watch?v=FIzXuLlynQE)


## License

The Monitoring Toolbox are licensed under the term of the CoreMedia Open Source License.
You will find a copy of this license in the LICENSE.md file included in the source package.


## Contributing

There are many ways to contribute to Monitoring Toolbox -- whether it be sending patches,
testing, reporting bugs, or reviewing and updating the documentation.

**Every contribution is appreciated!**

Please continue reading in the [contributing chapter](CONTRIBUTING.md).


<!-- TOC URLs -->
[About]: #about
[License]: #license
[Installation]: #installation
[Contributing]: #contributing
[Publications]: #publications



CHANGELOG.md

# Changelog

## 1810

- update to grafana 5.2.4
- update to icinga2 2.9.2
- remove postgres support in data-collector
- enhanced support of moebius' multi-node-deployment
  - import only the really needed Grafana templates
  - support multiple installations of CAE(Live) on one host
- better errorhandling at importing of grafana dashboards
- support for the Publisher mbean (used in the content-management-server application)
- add check_publisher for icinga2
- add open file descriptors monitoring to icinga and grafana
- fix double entries in the prepared service cache
- more measuring points for mysql / mariadb monitoring
- change markdown browser backend to algernon
- remove postgres support
- enable cors for external integration
- remove obsolete code


## 1808

- updates for mongodb 3.6
  - MongoDB 3.6 removes the deprecated HTTP interface and REST API to MongoDB
    we support now the official ruby client
- support alpine 3.8
- update to grafana 5.2.2
- update to icinga2 2.9.1
- update to icingaweb2 1.6.1
- update to jolokia 1.6.0
- updates for mongodb 3.6 (remove deprecated HTTP interface and REST API)
- updates for more than one node deployment
- enable the refresh task in the service-discovery per default
- fix bug with ecommerce cache dashboard
- fix some grafana 5 dashboards
- create team prefixed docker-compose files
- fix bug for content server detection when the hostname was 'localhost'


## 1806

- update to grafana 5.1.3
- grafana folder support
- set grafana 5 as default
- split annotations into more types
- update to icinga2 2.8.4
- update to icingaweb2 2.5.3
- update to jolokia 1.5.0
- update to graphite 1.1.3
- jolokia supports now authentication
- update to grafana gem 0.10.1
- update to icinga2 gem 1.0.0
- add new documentations container
  - including the stack documentation and a operation guide
- support node_exporter v0.16.x


## 1804

- add more documentations
- bugfix for icinga2 memory check
- update to grafana 5.0.4
- update icingaweb2
- set warning and critical values in some icinga2 checks
- add max feeder elements for CAE Feeders
- add this web UI
- add Background Feeder graphs for Content Feeder
- remove 'used UAPI Cache' from Grafana Overview


## 1803

- environments
  - update data-visualisation environment
  - update data-compose environment

- default log output is now stdout
- DNS resolver rewritten
- change output format
- singulary environment and json are now possible
- add Loglevel over Environment Variable
  * you can now set a LOG_LEVEL Environment to modify the Loglevel
    possible values are: DEBUG < INFO < WARN < ERROR < FATAL < UNKNOWN
    default is INFO
    no valid value will set to UNKNOWN
- remove some obsolete parts

- use new version for icinga2 certificate exchange (0.17.5)

- fix timing problems for RLS (#126)

- icinga2-master / icinga2-satellite / icinga-client
  - update the complete icinga suite
  - enable safe master/satellite communication with self-monitoring
  - use icinga2 gem in version 0.9.3.7 (1.0-rc)

- dashing
  - update for icinga-cert-service
  - use puma instead of thin

- grafana / grafana-client
  - overview dashboards can now be grouped by an array of tags
  - use the new annotation feature of Grafana 4.6
    remove the graphite-client, container is no longer needed
  - fix typos
  - fix bug with template_directory
  - enable grafana5
  - the grafana-client used now ruby erb templates insteed prepared json files

- carbon-client
  - code optimizations
  - add support for CoreMedia/cms#13125

- remove cm-external-discover container (is no longer needed)

- https support for webservice

- more documentations

## 1712


**BREAKING CHANGES**
 - The *internal API Port* was changed from 4567 to 8080
 - The Icinga2 Container has updated to v2.8.0 including certificateshandling
    **the docker volumes should be deleted for this container to create a new CA:** `docker volume rm -f aio_icinga`

----
update some containers
add healthchecks to all Containers
rewrite dokumentation
consolidate all Dockerfiles
reduce environment variables
move some mq-jobs from the rest-service to service-discovery

- data-collector
  add exception for missing MLS  DNS entry
  all non-coremedia services (mysql, node_exporter, etc.) can now be in the cm-service.yaml configured
  e.g. port and cretentials

      mysql:
        description: MySQL
        port: 3306
        monitoring_user: monitoring
        monitoring_password: monitoring

- service-discovery
  implement a 'rediscovery' feature
  after the service refreshing for the data-collector
  update the grafana dashboards.
  the overview and the license dashboard must be deleted before we can call the create_dashboard_for_host() function.

- icinga-client
  add certificate support for the icinga-client
  update cm-icinga-client to use the service refresh event from our service-discovery

- docker-compose
  remove all hard coded network ip
  use now the DNS from docker-host
  put all environment variables from environment.yaml back to docker-compose.yaml (possible depricated)

- grafana-client
  update grafana gem
  update all API calls
  integrate the new annotation feature of grafana (#104)
  the configure_server function for grafana can now handle inline ruby erb template parts

- rest-service
  minor improvements
  sanity checks for the REST API
  interanl cleanup and use of ruby styleguide

- icinga2
  add node_exporter check for load, memory and filesystem
  change icinga check for the sequence numbers between msl and rls
  use icinga2 side-channel to inject an inotify script for watching added hosts

- Dashboards:
  fix short link for mongodb
  add load, and cpu count for the OS row


## 1711

- recreate monitoring class as module and use a better structure
- optimize the rest service
- fix bug with the 'force' flag for creating hosts (#85)
- support https (#98)
- integrate ruby gem for grafana (#91)
- update some docker container
- enhance icinga2 checks (#84)

docker-cm-icinga-client/rootfs/init/icinga_cert.sh

ICINGA_API_PORT=${ICINGA_API_PORT:-5665}
USE_CERT_SERVICE=${USE_CERT_SERVICE:-'false'}

ICINGA_MASTER=${ICINGA_MASTER:-${ICINGA_HOST}}

# ICINGA_CERT_SERVICE_SERVER=
# ICINGA_CERT_SERVICE_PORT=
# ICINGA_CERT_SERVICE_BA_USER=
# ICINGA_CERT_SERVICE_BA_PASSWORD=
# ICINGA_CERT_SERVICE_API_USER=
# ICINGA_CERT_SERVICE_API_PASSWORD=

# get a new icinga certificate from our icinga-master
#
#
get_certificate() {

  validate_local_ca

  [[ -f ${WORK_DIR}/pki/${HOSTNAME}/${HOSTNAME}.key ]] && return

  if [[ "${USE_CERT_SERVICE}" == "true" ]]
  then
    log_info "we ask our cert-service for a certificate .."

    . /init/wait_for/cert_service.sh

    # generate a certificate request
    #
    code=$(curl \
      --user ${ICINGA_CERT_SERVICE_BA_USER}:${ICINGA_CERT_SERVICE_BA_PASSWORD} \
      --silent \
      --request GET \
      --header "X-API-USER: ${ICINGA_CERT_SERVICE_API_USER}" \
      --header "X-API-PASSWORD: ${ICINGA_CERT_SERVICE_API_PASSWORD}" \
      --write-out "%{http_code}\n" \
      --output /tmp/request_${HOSTNAME}.json \
      http://${ICINGA_CERT_SERVICE_SERVER}:${ICINGA_CERT_SERVICE_PORT}/${ICINGA_CERT_SERVICE_PATH}/v2/request/${HOSTNAME})

    if ( [[ $? -eq 0 ]] && [[ ${code} -eq 200 ]] )
    then
      sleep 4s

      log_info "certifiacte request was successful"
      log_info "download and install the certificate"

      master_name=$(jq --raw-output .master_name /tmp/request_${HOSTNAME}.json)
      checksum=$(jq --raw-output .checksum /tmp/request_${HOSTNAME}.json)

      mkdir -p ${WORK_DIR}/pki/${HOSTNAME}

      cp -a /tmp/request_${HOSTNAME}.json ${WORK_DIR}/pki/${HOSTNAME}/

      . /init/wait_for/cert_service.sh

      # get our created cert
      #
      code=$(curl \
        --user ${ICINGA_CERT_SERVICE_BA_USER}:${ICINGA_CERT_SERVICE_BA_PASSWORD} \
        --silent \
        --request GET \
        --header "X-API-USER: ${ICINGA_CERT_SERVICE_API_USER}" \
        --header "X-API-PASSWORD: ${ICINGA_CERT_SERVICE_API_PASSWORD}" \
        --header "X-CHECKSUM: ${checksum}" \
        --write-out "%{http_code}\n" \
        --request GET \
        --output ${WORK_DIR}/pki/${HOSTNAME}/${HOSTNAME}.tgz \
        http://${ICINGA_CERT_SERVICE_SERVER}:${ICINGA_CERT_SERVICE_PORT}/${ICINGA_CERT_SERVICE_PATH}/v2/cert/${HOSTNAME})

      if ( [[ $? -eq 0 ]] && [[ ${code} -eq 200 ]] )
      then
        cd ${WORK_DIR}/pki/${HOSTNAME}

        # the download has not working
        #
        if [[ ! -f ${HOSTNAME}.tgz ]]
        then
          log_error "Cert File '${HOSTNAME}.tgz' not found!"
          exit 1
        fi

        tar -xzf ${HOSTNAME}.tgz

        [[ ! -f ${HOSTNAME}.pem ]] && cat ${HOSTNAME}.crt ${HOSTNAME}.key >> ${HOSTNAME}.pem

        # store the master for later restart
        #
        echo "${master_name}" > ${WORK_DIR}/pki/${HOSTNAME}/master

        sleep 10s

        restart_master
      else
        log_error "can't download our certificate!"

        rm -rf ${WORK_DIR}/pki/${HOSTNAME} 2> /dev/null

        unset ICINGA_API_PKI_PATH

        exit 1
      fi
    else

      if [[ -f /tmp/request_${HOSTNAME}.json ]]
      then
        error=$(cat /tmp/request_${HOSTNAME}.json)

        log_error "${code} - the cert-service tell us a problem: '${error}'"
        log_error "exit ..."

        rm -f /tmp/request_${HOSTNAME}.json
      else
        log_error "${code} - the cert-service has an unknown problem."
      fi
      exit 1
    fi
  fi
}

# validate our lokal certificate against our certificate service
# with an API Request against
# http://${ICINGA_CERT_SERVICE_SERVER}:${ICINGA_CERT_SERVICE_PORT}${ICINGA_CERT_SERVICE_PATH}v2/validate/${checksum})
#
# if this failed, the PKI schould be removed
#
validate_local_ca() {

  if [[ -f ${WORK_DIR}/pki/${HOSTNAME}/ca.crt ]]
  then
    checksum=$(sha256sum ${WORK_DIR}/pki/${HOSTNAME}/ca.crt | cut -f 1 -d ' ')

    # validate our ca file
    #
    code=$(curl \
      --user ${ICINGA_CERT_SERVICE_BA_USER}:${ICINGA_CERT_SERVICE_BA_PASSWORD} \
      --silent \
      --request GET \
      --header "X-API-USER: ${ICINGA_CERT_SERVICE_API_USER}" \
      --header "X-API-PASSWORD: ${ICINGA_CERT_SERVICE_API_PASSWORD}" \
      --write-out "%{http_code}\n" \
      --output /tmp/validate_ca_${HOSTNAME}.json \
      http://${ICINGA_CERT_SERVICE_SERVER}:${ICINGA_CERT_SERVICE_PORT}/${ICINGA_CERT_SERVICE_PATH}/v2/validate/${checksum})

    if ( [[ $? -eq 0 ]] && [[ ${code} == 200 ]] )
    then
      rm -f /tmp/validate_ca_${HOSTNAME}.json
    else

      status=$(echo "${code}" | jq --raw-output .status 2> /dev/null)
      message=$(echo "${code}" | jq --raw-output .message 2> /dev/null)

      log_warn "our master has a new CA"

      rm -f /tmp/validate_ca_${HOSTNAME}.json
      rm -rf ${WORK_DIR}/pki
    fi
  else
    # we have no local cert file ..
    :
  fi
}

validate_cert() {

  if [[ -f ${WORK_DIR}/pki/${HOSTNAME}/${HOSTNAME}.pem ]]
  then
    log_info "validate our certifiacte"

    . /init/wait_for/icinga_master.sh

    code=$(curl \
      --silent \
      --insecure \
      --user ${ICINGA_CERT_SERVICE_API_USER}:${ICINGA_CERT_SERVICE_API_PASSWORD} \
      --capath ${WORK_DIR}/pki/${HOSTNAME} \
      --cert ${WORK_DIR}/pki/${HOSTNAME}/${HOSTNAME}.pem \
      --cacert ${WORK_DIR}/pki/${HOSTNAME}/ca.crt \
      https://${ICINGA_MASTER}:5665/v1/status/IcingaApplication)

    if [[ $? -eq 0 ]]
    then
      log_info "certifiacte is valid"
    else
      log_error ${code}
      log_error "certifiacte is invalid"
      log_info "unset PKI Variables to use Fallback"

      unset ICINGA_API_PKI_PATH
      unset ICINGA_API_NODE_NAME
    fi
  fi
}


# extract vars from the ICINGA_CERT_SERVICE environment variable
#
# we check first, if the value 'true' or 'false' for realize a backward compatibility
# if the value not empty, 'true' or 'false' we verify if this a valid json and extract our needed
# variables
# extract Variables from the json Environment
#
# example:
#   ICINGA_CERT_SERVICE: '{
#     "ba": { "user":"admin", "password":"admin" },
#     "api": { "user":"root", "password":"icinga" },
#     "server": "cm-icinga2-master",
#     "port": 4567,
#     "path": "/"
#   }'
#
#
extract_vars() {

  # default values for our Environment
  #
  ICINGA_CERT_SERVICE_SERVER=${ICINGA_CERT_SERVICE_SERVER:-}
  ICINGA_CERT_SERVICE_PORT=${ICINGA_CERT_SERVICE_PORT:-8080}
  ICINGA_CERT_SERVICE_PATH=${ICINGA_CERT_SERVICE_PATH:-'/'}
  ICINGA_CERT_SERVICE_API_USER=${ICINGA_CERT_SERVICE_API_USER:-'root'}
  ICINGA_CERT_SERVICE_API_PASSWORD=${ICINGA_CERT_SERVICE_API_PASSWORD:-'icinga'}
  ICINGA_CERT_SERVICE_BA_USER=${ICINGA_CERT_SERVICE_BA_USER:-"admin"}
  ICINGA_CERT_SERVICE_BA_PASSWORD=${ICINGA_CERT_SERVICE_BA_PASSWORD:-"admin"}

  USE_JSON="true"

  # detect if 'ICINGA_CERT_SERVICE' an json
  #
  if ( [[ ! -z "${ICINGA_CERT_SERVICE}" ]] && [[ "${ICINGA_CERT_SERVICE}" != "true" ]] && [[ "${ICINGA_CERT_SERVICE}" != "false" ]] )
  then
    echo "${ICINGA_CERT_SERVICE}" | json_verify -q 2> /dev/null

    if [[ $? -gt 0 ]]
    then
      log_info "the ICINGA_CERT_SERVICE Environment is not an json"
#       log_info "use fallback strategy."
      USE_JSON="false"
    fi
  else
    log_info "the ICINGA_CERT_SERVICE Environment is not an json"
#     log_info "use fallback strategy."
    USE_JSON="false"
  fi

  # we can use json as configure
  #
  if [[ "${USE_JSON}" == "true" ]]
  then

    log_info "the ICINGA_CERT_SERVICE Environment is an json"

    if ( [[ "${ICINGA_CERT_SERVICE}" == "true" ]] || [[ "${ICINGA_CERT_SERVICE}" == "false" ]] )
    then
      log_error "the ICINGA_CERT_SERVICE Environment must be an json, not true or false!"
    else
      ICINGA_CERT_SERVICE_SERVER=$(echo "${ICINGA_CERT_SERVICE}" | jq --raw-output .server)
      ICINGA_CERT_SERVICE_PORT=$(echo "${ICINGA_CERT_SERVICE}" | jq --raw-output .port)
      ICINGA_CERT_SERVICE_PATH=$(echo "${ICINGA_CERT_SERVICE}" | jq --raw-output .path)
      ICINGA_CERT_SERVICE_API_USER=$(echo "${ICINGA_CERT_SERVICE}" | jq --raw-output .api.user)
      ICINGA_CERT_SERVICE_API_PASSWORD=$(echo "${ICINGA_CERT_SERVICE}" | jq --raw-output .api.password)
      ICINGA_CERT_SERVICE_BA_USER=$(echo "${ICINGA_CERT_SERVICE}" | jq --raw-output .ba.user)
      ICINGA_CERT_SERVICE_BA_PASSWORD=$(echo "${ICINGA_CERT_SERVICE}" | jq --raw-output .ba.password)

      [[ "${ICINGA_CERT_SERVICE_SERVER}" == null ]] && ICINGA_CERT_SERVICE_SERVER=
      [[ "${ICINGA_CERT_SERVICE_PORT}" == null ]] && ICINGA_CERT_SERVICE_PORT=8080
      [[ "${ICINGA_CERT_SERVICE_PATH}" == null ]] && ICINGA_CERT_SERVICE_PATH="/"
      [[ "${ICINGA_CERT_SERVICE_API_USER}" == null ]] && ICINGA_CERT_SERVICE_API_USER=
      [[ "${ICINGA_CERT_SERVICE_API_PASSWORD}" == null ]] && ICINGA_CERT_SERVICE_API_PASSWORD=
      [[ "${ICINGA_CERT_SERVICE_BA_USER}" == null ]] && ICINGA_CERT_SERVICE_BA_USER=
      [[ "${ICINGA_CERT_SERVICE_BA_PASSWORD}" == null ]] && ICINGA_CERT_SERVICE_BA_PASSWORD=
    fi
  else

    ICINGA_CERT_SERVICE_SERVER=${ICINGA_CERT_SERVICE_SERVER:-}
    ICINGA_CERT_SERVICE_PORT=${ICINGA_CERT_SERVICE_PORT:-8080}
    ICINGA_CERT_SERVICE_PATH=${ICINGA_CERT_SERVICE_PATH:-'/'}
    ICINGA_CERT_SERVICE_API_USER=${ICINGA_CERT_SERVICE_API_USER:-''}
    ICINGA_CERT_SERVICE_API_PASSWORD=${ICINGA_CERT_SERVICE_API_PASSWORD:-''}
    ICINGA_CERT_SERVICE_BA_USER=${ICINGA_CERT_SERVICE_BA_USER:-"admin"}
    ICINGA_CERT_SERVICE_BA_PASSWORD=${ICINGA_CERT_SERVICE_BA_PASSWORD:-"admin"}
  fi

  ICINGA_CERT_SERVICE_PATH=$(echo "${ICINGA_CERT_SERVICE_PATH}" | cut -d "/" -f 2)

  validate_certservice_environment
}


validate_certservice_environment() {

  ICINGA_CERT_SERVICE_BA_USER=${ICINGA_CERT_SERVICE_BA_USER:-"admin"}
  ICINGA_CERT_SERVICE_BA_PASSWORD=${ICINGA_CERT_SERVICE_BA_PASSWORD:-"admin"}
  ICINGA_CERT_SERVICE_API_USER=${ICINGA_CERT_SERVICE_API_USER:-""}
  ICINGA_CERT_SERVICE_API_PASSWORD=${ICINGA_CERT_SERVICE_API_PASSWORD:-""}
  ICINGA_CERT_SERVICE_SERVER=${ICINGA_CERT_SERVICE_SERVER:-"localhost"}
  ICINGA_CERT_SERVICE_PORT=${ICINGA_CERT_SERVICE_PORT:-"80"}
  ICINGA_CERT_SERVICE_PATH=${ICINGA_CERT_SERVICE_PATH:-"/"}
  USE_CERT_SERVICE=false

  # use the new Cert Service to create and get a valide certificat for distributed icinga services
  #
  if (
    [[ ! -z ${ICINGA_CERT_SERVICE_SERVER} ]] &&
    [[ ! -z ${ICINGA_CERT_SERVICE_BA_USER} ]] &&
    [[ ! -z ${ICINGA_CERT_SERVICE_BA_PASSWORD} ]] &&
    [[ ! -z ${ICINGA_CERT_SERVICE_API_USER} ]] &&
    [[ ! -z ${ICINGA_CERT_SERVICE_API_PASSWORD} ]]
  )
  then
    USE_CERT_SERVICE=true

    export ICINGA_CERT_SERVICE_BA_USER
    export ICINGA_CERT_SERVICE_BA_PASSWORD
    export ICINGA_CERT_SERVICE_API_USER
    export ICINGA_CERT_SERVICE_API_PASSWORD
    export ICINGA_CERT_SERVICE_SERVER
    export ICINGA_CERT_SERVICE_PORT
    export ICINGA_CERT_SERVICE_PATH
    export USE_CERT_SERVICE
  fi
}


restart_master() {

  sleep $(shuf -i 5-30 -n 1)s

  . /init/wait_for/icinga_master.sh

  # restart the master to activate the zone
  #
  log_info "restart the master '${ICINGA_MASTER}' to activate our certificate"
  code=$(curl \
    --user ${ICINGA_CERT_SERVICE_API_USER}:${ICINGA_CERT_SERVICE_API_PASSWORD} \
    --silent \
    --header 'Accept: application/json' \
    --request POST \
    --insecure \
    https://${ICINGA_MASTER}:5665/v1/actions/restart-process )

  if [[ $? -gt 0 ]]
  then
    status=$(echo "${code}" | jq --raw-output '.results[].code' 2> /dev/null)
    message=$(echo "${code}" | jq --raw-output '.results[].status' 2> /dev/null)

    log_error "${code}"
    log_error "${message}"
  fi

  sleep 5s
}

if [[ -z ${ICINGA_MASTER} ]]
then
  log_error "the ICINGA_MASTER environment is missing."
  exit 1
fi


extract_vars
. /init/wait_for/cert_service.sh
get_certificate
validate_cert

if [[ -d ${WORK_DIR}/pki/${HOSTNAME} ]]
then
  log_info "export PKI vars"

  export ICINGA_HOST=${ICINGA_MASTER}

  export ICINGA_API_USER=${ICINGA_CERT_SERVICE_API_USER}
  export ICINGA_API_PASSWORD=${ICINGA_CERT_SERVICE_API_PASSWORD}

  export ICINGA_API_PKI_PATH=${WORK_DIR}/pki/${HOSTNAME}
  export ICINGA_API_NODE_NAME=${HOSTNAME}
fi
